network/cluster.go

```diff
@@ -0,0 +1,267 @@
+	agentCont, exists := c.agents[id]
+	if !exists {
+		panic(fmt.Sprintf("Agent does not exist: %d", id))
+	}
+	return agentCont
+}
+
+func (c *Cluster) RemoveAgent(id int) error {
+	agent := c.GetAgent(id)
+	err := agent.Stop()
+	if err != nil {
+		return err
+	}
+
+	c.lock.Lock()
+	defer c.lock.Unlock()
+	delete(c.agents, id)
+	return nil
+}
+
+func (c *Cluster) CreateConn(baseConn net.Conn, leftPortID, rightPortID int) (net.Conn, error) {
+	lcity, rcity := c.GetAgentCity(leftPortID), c.GetAgentCity(rightPortID)
+	latencyDuration := c.latencyFinder(lcity, rcity)
+
+	nn := lat.Network{
+		Kbps:    getValue(c.config.Kbps, defaultKbps),
+		Latency: latencyDuration,
+		MTU:     getValue(c.config.MTU, defaultMTU),
+	}
+	return nn.Conn(baseConn)
+}
+
+func (c *Cluster) Gossip(fromID int, size int) {
+	agent := c.GetAgent(fromID)
+	buf := generateMsg(c.config.MsgSize)
+	agent.Gossip(buf)
+}
+
+func (c *Cluster) Connect(fromID, toID int) error {
+	from, to := c.GetAgent(fromID), c.GetAgent(toID)
+	if from.NumPeers() == c.config.MaxPeers {
+		return fmt.Errorf("agent %d has already maximum peers connected", fromID)
+	}
+	return from.Connect(to)
+}
+
+func (c *Cluster) Stop(portID int) error {
+	return c.GetAgent(portID).Stop()
+}
+
+func (c *Cluster) Start(portID int) error {
+	return c.GetAgent(portID).Listen(c.config.Ip, portID)
+}
+
+func (c *Cluster) StopAll() {
+	for _, cont := range c.agents {
+		cont.agent.Stop()
+	}
+}
+
+func (c *Cluster) GossipLoop(context context.Context, gossipTime time.Duration, timeout time.Duration) (int64, int64) {
+	msgsPublishedCnt, msgsFailedCnt := int64(0), int64(0)
+	ch := make(chan struct{})
+	for _, cont := range c.agents {
+		go func(a ClusterAgent) {
+			tm := time.NewTicker(gossipTime)
+			defer tm.Stop()
+		outer:
+			for {
+				select {
+				case <-tm.C:
+					err := a.Gossip(generateMsg(c.config.MsgSize))
+					if err == nil {
+						atomic.AddInt64(&msgsPublishedCnt, 1)
+					} else {
+						atomic.AddInt64(&msgsFailedCnt, 1)
+					}
+				case <-ch:
+					break outer
+				}
+			}
+
+		}(cont.agent)
+	}
+
+	select {
+	case <-time.After(timeout):
+	case <-context.Done():
+	}
+
+	close(ch)
+	return msgsPublishedCnt, msgsFailedCnt
+}
+
+func (c *Cluster) MsgReceived(lid, rid string, data []byte) {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+	if c.agentStats[lid] == nil {
+		c.agentStats[lid] = make(map[string]int64)
+	}
+	c.agentStats[lid][rid]++
+}
+
+func (c *Cluster) PrintReceiversStats() {
+	all := int64(0)
+	byAgent := make(map[string]int64, 0)
+	for k, v := range c.agentStats {
+		cnt := int64(0)
+		for _, num := range v {
+			all += num
+			cnt += num
+		}
+		byAgent[k] = cnt
+	}
+
+	fmt.Printf("Recieved messages: %d\n", all)
+	fmt.Println("Recieved messages by agent")
+	for k, v := range byAgent {
+		fmt.Printf("Recieved messages by %s: %d\n", k, v)
+	}
+}
+
+func (c *Cluster) StartAgents(agentsNumber, routinesNumber int, factory func(id int) (ClusterAgent, string)) (int64, time.Duration) {
```

Comment:
We can hardcode internally the `routinesNumber` number to 5 for now. This way we make `StartAgents` much more simple.

Reply:
Fixed